File: Binary_Search /python/2805.py
n,m = map(int,input().split())
nums = list(map(int,input().split()))
result = 0
def search(min_num,max_num):
  global result

  if min_num>max_num: return
  half_num = min_num + (max_num-min_num)//2
  height = 0
  for num in nums:
    temp = num-half_num
    if (temp>0): height+=temp
  if height>=m:
    result = half_num
    search(half_num+1,max_num)
  else:
    search(min_num,half_num-1)
search(0,max(nums))
print(result)

File: Binary_Search /javascript/2805.js
const input = require('fs').readFileSync('/dev/stdin').toString().split('\n');

const search = (minNum,maxNum) => {
  if (minNum > maxNum) return;
  let halfNum = minNum + Math.floor((maxNum-minNum)/2);
  let dif, height = 0;

  for (let i = 0 ;i<nums.length ; i++){
    dif = nums[i]-halfNum;
    if (dif>0) height+=dif;
  }
  if (height >= M){
    result = halfNum;
    search(halfNum+1,maxNum);
  }
  else{
    search(minNum,halfNum-1);
  }
}


const [N,M] = input[0].split(' ');
const nums = input[1].split(' ').map(Number);
let result = 0;
const maxHeight = nums.reduce((prev,cur)=>{
  if(prev<cur) prev = cur;
  return prev;
},0)
search(0,maxHeight);
console.log(result);